File: src/geo_search.R
# This script queries the NCBI GEO database for the number of gene expression microarray and 
# sequencing datasets for a given species each year.

# Output:
# "result/per_year_per_type_geo_for_{species}.pdf"
# "result/per_year_array_v_seq_for_{species}.pdf"
# "result/geo_query_result_for_{species}.tsv"

# 0. Prep -----------------------------------------------------------------

library(here)
library(rentrez)
library(tidyverse)
library(ggh4x)

theme_set(cowplot::theme_minimal_hgrid())

# Define two functions.
# Function to query NCBI GEO for given parameters. I want the number of studies each year.
query_ncbi_geo <- function(term, species, start_year, end_year) {
    results <- map_dfr(start_year:end_year, function(year) {
        query <- paste(term, "[DataSet Type]", "AND", year, "[PDAT]", "AND", species, "[ORGN]", sep = " ")
        search <- entrez_search(db = "gds", term = query)
        tibble(Year = year, Count = search$count, DataSetType = term)
    })
    return(results)
}

# Function to reorder factor levels based on total count
order_facet_by_total_count <- function(data) {
    total_counts <- data %>%
        group_by(DataSetType) %>%
        summarise(TotalCount = sum(Count)) %>%
        arrange(desc(TotalCount))
    data %>%
        mutate(DataSetType = factor(DataSetType, levels = total_counts$DataSetType))
}

# 1. Set up query terms ---------------------------------------------------

# Set inputs
start_year <- 2000
end_year <- 2023
species <- "Zea mays" # Arabidopsis thaliana, Oryza sativa, Zea mays, Homo sapiens

# Define the terms for gene expression microarray and Sequencing
terms <- list(
    microarray = c(
        "expression profiling by array",
        "expression profiling by genome tiling array",
        "expression profiling by snp array",
        "genome binding/occupancy profiling by array",
        "genome binding/occupancy profiling by genome tiling array",
        "genome binding/occupancy profiling by snp array",
        "genome variation profiling by array",
        "genome variation profiling by genome tiling array",
        "genome variation profiling by snp array",
        "methylation profiling by array",
        "methylation profiling by genome tiling array",
        "methylation profiling by snp array",
        "non coding rna profiling by array",
        "non coding rna profiling by genome tiling array",
        "protein profiling by protein array",
        "snp genotyping by snp array"
    ),
    sequencing = c(
        "expression profiling by high throughput sequencing",
        "genome binding/occupancy profiling by high throughput sequencing",
        "genome variation profiling by high throughput sequencing",
        "methylation profiling by high throughput sequencing",
        "non coding rna profiling by high throughput sequencing"
    )
)

# 2. Query the NCBI -------------------------------------------------------

# Query NCBI GEO for gene expression microarray and sequencing data
result_all <- map_dfr(names(terms), function(type) {
    map_dfr(terms[[type]], query_ncbi_geo, species, start_year, end_year) %>%
        mutate(Type = type)
})

# Reorder the DataSetType factor based on total counts
result_all_order <- order_facet_by_total_count(result_all)

# Summarize data by year and type
year_sum <- result_all_order %>%
    group_by(Year, Type) %>%
    summarise(TotalCount = sum(Count))

# 3. Plot -----------------------------------------------------------------

# Plotting microarray data
p_mic <- ggplot(result_all_order %>% filter(Type == "microarray"), aes(x = Year, y = Count)) +
    geom_point() +
    geom_smooth(se = FALSE, color = "#0072B2") +
    ylab("Number of GEO Datasets") +
    ggtitle(paste0("Number of array datasets in GEO for ", species)) +
    facet_wrap2(~ DataSetType, scales = "free_y", nrow = 4, ncol = 4, trim_blank = FALSE) +
    theme(strip.background = element_rect(fill = "grey80"),
          strip.text = element_text(color = "black", hjust = 0))

# Plotting sequencing data
p_seq <- ggplot(result_all_order %>% filter(Type == "sequencing"), aes(x = Year, y = Count)) +
    geom_point() +
    geom_smooth(se = FALSE, color = "#D55E00") +
    ylab("Number of GEO Datasets") +
    ggtitle(paste0("Number of sequencing datasets in GEO for ", species)) +
    facet_wrap2(~ DataSetType, scales = "free_y", nrow = 4, ncol = 4, trim_blank = FALSE) +
    theme(strip.background = element_rect(fill = "grey80"),
          strip.text = element_text(color = "black", hjust = 0))

# Plotting the total number of GEO datasets each year
p_year <- ggplot(year_sum, aes(Year, TotalCount, fill = Type)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    scale_fill_manual(values = c("microarray" = "#0072B2", 
                                 "sequencing" = "#D55E00")) +
    labs(y = "Number of GEO datasets each year",
         title = paste0("Total number of GEO datasets for ", species))

# 4. Save plots and result table ------------------------------------------

pdf(here("result", paste0("per_year_per_type_geo_for_", gsub(" ", "", species), ".pdf")),
    width = 12, height = 8)
print(p_mic) 
print(p_seq)
dev.off()

pdf(here("result", paste0("per_year_array_v_seq_for_", gsub(" ", "", species), ".pdf")),
    width = 6, height = 4)
print(p_year)
dev.off()

write_tsv(result_all_order, 
          file = here("result",
                      paste0("geo_query_result_for_", gsub(" ", "", species), ".tsv")))

File: src/geo_search.py
import polars as pl
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from Bio import Entrez

Entrez.email = "xyz@email.com"


def query_ncbi_geo(term, species, start_year, end_year):
    results = []
    for year in range(start_year, end_year + 1):
        query = f"{term} [DataSet Type] AND {year} [PDAT] AND {species} [ORGN]"
        search = Entrez.esearch(db="gds", term=query)
        record = Entrez.read(search)
        results.append(
            {"Year": year, "Count": int(record["Count"]), "DataSetType": term}
        )
    return pl.DataFrame(results)


def order_by_total_count(data):
    total_counts = (
        data.groupby("DataSetType")
        .agg(pl.col("Count").sum().alias("TotalCount"))
        .sort("TotalCount", reverse=True)
    )
    ordered_categories = total_counts["DataSetType"]
    return data.with_column(
        pl.col("DataSetType").cast(pl.Categorical).cat.set_ordering(ordered_categories)
    )


terms = {
    "microarray": [
        "expression profiling by array",
        "expression profiling by genome tiling array",
        "expression profiling by snp array",
        "genome binding/occupancy profiling by array",
        "genome binding/occupancy profiling by genome tiling array",
        "genome binding/occupancy profiling by snp array",
        "genome variation profiling by array",
        "genome variation profiling by genome tiling array",
        "genome variation profiling by snp array",
        "methylation profiling by array",
        "methylation profiling by genome tiling array",
        "methylation profiling by snp array",
        "non coding rna profiling by array",
        "non coding rna profiling by genome tiling array",
        "protein profiling by protein array",
        "snp genotyping by snp array",
    ],
    "sequencing": [
        "expression profiling by high throughput sequencing",
        "genome binding/occupancy profiling by high throughput sequencing",
        "genome variation profiling by high throughput sequencing",
        "methylation profiling by high throughput sequencing",
        "non coding rna profiling by high throughput sequencing",
    ],
}
species = "Zea mays"
start_year = 2022
end_year = 2023

results = [
    query_ncbi_geo(term, species, start_year, end_year)
    for type_, terms_list in terms.items()
    for term in terms_list
]

results = []
for type_, terms_list in terms.items():
    for term in terms_list:
        result_df = query_ncbi_geo(term, species, start_year, end_year)
        results.append(result_df)
list(terms.values())[0][1]
list(terms.keys())
next(iter(terms.keys()))

result_all = pl.concat(results)
tt2 = result_all.to_pandas()

tt2.dtypes
pl.Enum(tt2_order)
cat_type = pd.CategoricalDtype(categories=tt2_order, ordered=True)
# Convert the string column to an ordered categorical column
tt2["DataSetType"] = tt2["DataSetType"].astype(cat_type)


print(tt2)
print(tt2["DataSetType"].dtype)  # Check the dtype to confirm it's an ordered categorical


# Directly order by total counts within the DataFrame
total_counts = result_all.groupby('DataSetType').agg(pl.col('Count').sum().alias('TotalCount'))

item_order = total_counts.sort("TotalCount", descending=True)
tt1 = total_counts.sort("TotalCount", descending=True)
tt1.to_series(0).to_list()
tt2_order = tt1.to_series().to_list()

df = result_all.with_columns(
    pl.col("DataSetType").cast(pl.Categorical).cat.set_ordering(tt2_order)
)


result_all_ordered = result_all.select(pl.col("DataSetType").cast(pl.Categorical)
)



result_all_ordered = order_by_total_count(result_all)

yearly_summary = (
    result_all.groupby(["Year", "DataSetType"])
    .agg(pl.col("Count").sum().alias("TotalCount"))
    .to_pandas()
)

sns.set(style="whitegrid")
plt.figure(figsize=(14, 7))
sns.barplot(data=yearly_summary, x="Year", y="TotalCount", hue="DataSetType")
plt.title(f"Annual Distribution of GEO Dataset Types for {species}")
plt.ylabel("Number of Datasets")
plt.xlabel("Year")
plt.legend(title="Dataset Type")
plt.tight_layout()
plt.show()
